File: app/src/main/java/com/jarvis/autocontroller/JarvisAutomationService.java
package com.jarvis.autocontroller;

import android.accessibilityservice.AccessibilityService;
import android.accessibilityservice.GestureDescription;
import android.content.SharedPreferences;
import android.graphics.Path;
import android.os.Build;
import android.view.accessibility.AccessibilityEvent;
import android.view.accessibility.AccessibilityNodeInfo;
import java.util.HashSet;
import java.util.Set;

public class JarvisAutomationService extends AccessibilityService {

    private static JarvisAutomationService instance;

    @Override
    public void onServiceConnected() {
        super.onServiceConnected();
        instance = this;
    }

    public static JarvisAutomationService getInstance() {
        return instance;
    }

    @Override
    public void onAccessibilityEvent(AccessibilityEvent event) {
        if (event == null || event.getPackageName() == null) return;
        String currentPackage = event.getPackageName().toString();

        SharedPreferences prefs = getSharedPreferences("JarvisSettings", MODE_PRIVATE);
        Set<String> selectedApps = prefs.getStringSet("selected_apps", new HashSet<String>());

        // ইউজার সেটিংসে সিলেক্ট করা অ্যাপ চেক করা
        if (selectedApps != null && selectedApps.contains(currentPackage)) {
            // উইন্ডোর কন্টেন্ট পরিবর্তন হলে নোড রিড করা বা অটো কাজ চালানো সম্ভব
            AccessibilityNodeInfo rootNode = getRootInActiveWindow();
            if (rootNode != null) {
                // এক্সাম্পল: নির্দিষ্ট কোনো বাটনে ক্লিক বা কাজ
                rootNode.recycle();
            }
        }
    }

    @Override
    public void onInterrupt() {}

    // স্ক্রিনের যেকোনো কোঅর্ডিনেটে অটো ট্যাপ করার মেথড
    public void autoClick(float x, float y) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            Path clickPath = new Path();
            clickPath.moveTo(x, y);
            GestureDescription.StrokeDescription clickStroke =
                    new GestureDescription.StrokeDescription(clickPath, 0, 50);
            GestureDescription.Builder gestureBuilder = new GestureDescription.Builder();
            gestureBuilder.addStroke(clickStroke);
            dispatchGesture(gestureBuilder.build(), null, null);
        }
    }

    // স্ক্রল ডাউন অটোমেশন মেথড
    public void autoScrollDown() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            Path scrollPath = new Path();
            // স্ক্রিনের মাঝখান থেকে নিচ থেকে উপরে সোয়াইপ (স্ক্রল ডাউন)
            scrollPath.moveTo(500, 1500);
            scrollPath.lineTo(500, 500);
            GestureDescription.StrokeDescription scrollStroke =
                    new GestureDescription.StrokeDescription(scrollPath, 0, 300);
            GestureDescription.Builder gestureBuilder = new GestureDescription.Builder();
            gestureBuilder.addStroke(scrollStroke);
            dispatchGesture(gestureBuilder.build(), null, null);
        }
    }

    // নোড টেক্সট বা আইডি দিয়ে ক্লিক
    public void clickByText(String text) {
        AccessibilityNodeInfo rootNode = getRootInActiveWindow();
        if (rootNode != null) {
            for (AccessibilityNodeInfo node : rootNode.findAccessibilityNodeInfosByText(text)) {
                node.performAction(AccessibilityNodeInfo.ACTION_CLICK);
            }
            rootNode.recycle();
        }
    }
}

File: app/src/main/java/com/example/JarvisAutomationService.kt
package com.example

import android.accessibilityservice.AccessibilityService
import android.accessibilityservice.GestureDescription
import android.content.Context
import android.content.SharedPreferences
import android.graphics.Path
import android.os.Build
import android.view.accessibility.AccessibilityEvent
import android.view.accessibility.AccessibilityNodeInfo
import java.util.HashSet

class JarvisAutomationService : AccessibilityService() {

    val automator: com.example.automation.EnhancedAccessibilityAutomator by lazy {
        com.example.automation.EnhancedAccessibilityAutomator { this }
    }

    companion object {
        @Volatile
        private var instance: JarvisAutomationService? = null

        @JvmStatic
        fun getInstance(): JarvisAutomationService? {
            return instance
        }

        fun getAutomator(): com.example.automation.EnhancedAccessibilityAutomator? {
            return instance?.automator
        }

        fun isServiceRunning(): Boolean {
            return instance != null
        }
    }

    override fun onServiceConnected() {
        super.onServiceConnected()
        instance = this
    }

    override fun onUnbind(intent: android.content.Intent?): Boolean {
        if (instance == this) {
            instance = null
        }
        return super.onUnbind(intent)
    }

    override fun onDestroy() {
        super.onDestroy()
        if (instance == this) {
            instance = null
        }
    }

    override fun onAccessibilityEvent(event: AccessibilityEvent?) {
        if (event == null || event.packageName == null) return
        val currentPackage = event.packageName.toString()

        try {
            // [HEAVY APP LOCKER] Intercept opening of protected applications with real overlay
            if (event.eventType == AccessibilityEvent.TYPE_WINDOW_STATE_CHANGED) {
                if (currentPackage != packageName && !com.example.security.JarvisAppLockOverlayActivity.isUnlocked(currentPackage)) {
                    val encHelper = com.example.security.EncryptedPrefsHelper(this)
                    val lockedApps = encHelper.getLockedApps()
                    if (lockedApps.contains(currentPackage)) {
                        val overlayIntent = android.content.Intent(this, com.example.security.JarvisAppLockOverlayActivity::class.java).apply {
                            flags = android.content.Intent.FLAG_ACTIVITY_NEW_TASK or android.content.Intent.FLAG_ACTIVITY_CLEAR_TOP
                            putExtra("TARGET_PACKAGE", currentPackage)
                        }
                        startActivity(overlayIntent)
                    }
                }
            }

            val prefs: SharedPreferences = getSharedPreferences("JarvisSettings", Context.MODE_PRIVATE)
            val selectedApps: Set<String> = prefs.getStringSet("selected_apps", HashSet<String>()) ?: HashSet<String>()

            // Inspect selected apps
            if (selectedApps.contains(currentPackage)) {
                val rootNode: AccessibilityNodeInfo? = try {
                    rootInActiveWindow
                } catch (_: Exception) {
                    null
                }
                rootNode?.recycle()
            }
        } catch (e: Exception) {
            // Safely guard against window transition and lifecycle anomalies
            e.printStackTrace()
        }
    }

    override fun onInterrupt() {
        // Handled cleanly when user disables service in settings
    }

    // স্ক্রিনের যেকোনো কোঅর্ডিনেটে অটো ট্যাপ করার মেথড
    fun autoClick(x: Float, y: Float): Boolean {
        return try {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                val clickPath = Path()
                clickPath.moveTo(x, y)
                val clickStroke = GestureDescription.StrokeDescription(clickPath, 0, 50)
                val gestureBuilder = GestureDescription.Builder()
                gestureBuilder.addStroke(clickStroke)
                dispatchGesture(gestureBuilder.build(), null, null)
            } else {
                false
            }
        } catch (e: Exception) {
            false
        }
    }

    // স্ক্রল ডাউন অটোমেশন মেথড
    fun autoScrollDown(): Boolean {
        return try {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                val scrollPath = Path()
                scrollPath.moveTo(500f, 1500f)
                scrollPath.lineTo(500f, 500f)
                val scrollStroke = GestureDescription.StrokeDescription(scrollPath, 0, 300)
                val gestureBuilder = GestureDescription.Builder()
                gestureBuilder.addStroke(scrollStroke)
                dispatchGesture(gestureBuilder.build(), null, null)
            } else {
                false
            }
        } catch (e: Exception) {
            false
        }
    }

    // নোড টেক্সট বা আইডি দিয়ে ক্লিক
    fun clickByText(text: String): Int {
        var clickedCount = 0
        try {
            val rootNode: AccessibilityNodeInfo? = try {
                rootInActiveWindow
            } catch (_: Exception) {
                null
            }
            if (rootNode != null) {
                val nodes = rootNode.findAccessibilityNodeInfosByText(text)
                if (nodes != null) {
                    for (node in nodes) {
                        try {
                            if (node.performAction(AccessibilityNodeInfo.ACTION_CLICK)) {
                                clickedCount++
                            }
                        } catch (_: Exception) {}
                    }
                }
                try {
                    rootNode.recycle()
                } catch (_: Exception) {}
            }
        } catch (e: Exception) {
            // Guarded
        }
        return clickedCount
    }

    // Extended UI Automation Helpers
    fun swipeHorizontal(fromLeftToRight: Boolean): Boolean {
        return if (fromLeftToRight) {
            automator.performSwipe(150f, 1000f, 900f, 1000f, 300L)
        } else {
            automator.performSwipe(900f, 1000f, 150f, 1000f, 300L)
        }
    }

    fun inputTextToField(targetQuery: String, text: String): Boolean {
        return automator.findAndSetText(targetQuery, text)
    }

    fun pressBack(): Boolean = automator.performGlobalBack()
    fun pressHome(): Boolean = automator.performGlobalHome()
    fun pressRecents(): Boolean = automator.performGlobalRecents()
}
